### PR TITLE
Add branch protection option to block merge on requested changes.

### LINKED ISSUE #9130: Block pull request merge if changes are requested
- Gitea version (or commit ref): 1.11.0+dev-311-g675f27523

## Description
There should be a branch protection setting to _Block merge if changes are requested by a reviewer_. A reviewer here means someone that would count to the required approvals (with current implementation that is the same as users in the whitelist).

For discussion: I would actually think blocking should be the default behavior and instead you can change the setting to _Allow merging even if there are changes requested_.

### PR DESCRIPTION
**Pull request by davidsvantesson**
Fix #9130.

## Screenshot
image

### CODE REVIEW

File: models/branches.go [hunk 1]
```diff
@@ -44,6 +44,7 @@ type ProtectedBranch struct {
 	ApprovalsWhitelistUserIDs []int64            `xorm:"JSON TEXT"`
 	ApprovalsWhitelistTeamIDs []int64            `xorm:"JSON TEXT"`
 	RequiredApprovals         int64              `xorm:"NOT NULL DEFAULT 0"`
+	BlockOnRejectedReviews    bool               `xorm:"NOT NULL DEFAULT false"`
```
Comment (techknowlogick) [inline models/branches.go:47]: Please add new migration for this column

File: options/locale/locale_en-US.ini [hunk 2]
```diff
@@ -1417,6 +1418,8 @@ settings.update_protect_branch_success = Branch protection for branch '%s' has b
 settings.remove_protected_branch_success = Branch protection for branch '%s' has been disabled.
 settings.protected_branch_deletion = Disable Branch Protection
 settings.protected_branch_deletion_desc = Disabling branch protection allows users with write permission to push to the branch. Continue?
+settings.block_rejected_reviews = Block merge on rejcted reviews
```
Comment (jolheiser) [inline options/locale/locale_en-US.ini]:
```suggestion
settings.block_rejected_reviews = Block merge on rejected reviews
```

File: options/locale/locale_en-US.ini [hunk 3]
```diff
@@ -1054,6 +1054,7 @@ pulls.is_checking = "Merge conflict checking is in progress. Try again in few mo
 pulls.required_status_check_failed = Some required checks were not successful.
 pulls.required_status_check_administrator = As an administrator, you may still merge this pull request.
 pulls.blocked_by_approvals = "This Pull Request doesn't have enough approvals yet. %d of %d approvals granted."
+pulls.blocked_by_rejection = "This Pull Request has enough approvals but is blocked by changes requested by an official reviewer."
```
Comment (jolheiser) [inline options/locale/locale_en-US.ini]:
```suggestion
pulls.blocked_by_rejection = "This Pull Request has changes requested by an official reviewer."
```

File: routers/private/hook.go [hunk 4]
```diff
@@ -113,6 +113,13 @@ func HookPreReceive(ctx *macaron.Context, opts private.HookOptions) {
 					})
 					return
 				}
+				if !protectBranch.MergeBlockedByRejectedReview(pr) {
```
Comment (jolheiser) [inline routers/private/hook.go]: Shouldn't this if-check be checking for `true`?

**Review by jolheiser — CHANGES_REQUESTED** [review]

File: models/branches.go (hunk 1 above)
Comment (davidsvantesson) [reply models/branches.go:47]: Done

File: routers/private/hook.go (hunk 4 above)
Comment (davidsvantesson) [reply routers/private/hook.go]: Test cases say that also 😅

File: models/migrations/v117.go [hunk 5]
```diff
@@ -0,0 +1,17 @@
+// Copyright 2019 The Gitea Authors. All rights reserved.
```
Comment (techknowlogick) [inline models/migrations/v117.go]:
```suggestion
// Copyright 2020 The Gitea Authors. All rights reserved.
```

Comment (davidsvantesson) [reply models/migrations/v117.go]: Right, forgot that 😄

**Comment by bagasme** [comment]
@davidsvantesson Supposed that `test` user has push new commits that address `test3`'s requested change review, although the review doesn't have any suggested changes, is there a way to pass it?

**Comment by davidsvantesson** [comment]
@bagasme I am not sure what you mean. If `test3` has requested changes that user then has to approve the PR to be able to merge it.

**Review by techknowlogick — APPROVED** [review]

**Review by jolheiser — APPROVED** [review]

**Review by lafriks — APPROVED** [review]

**Pull request merged**